### ISSUE TITLE
Require net YAML file in all frameworks

### ISSUE THREAD

**Issue by dilpath** [sebapersson/petab_sciml#17]
The net YAML file is not capable of representing all architectures currently. However, we can still require that all models have an accompanying net YAML file. For example, we can allow that the `forward` part of a net YAML is actually a script.
```yaml
mlmodel_id: model1
inputs:
- input_id: input1
layers:
- layer_id: layer1
  layer_type: Linear
  args: ...
- layer_id: layer2
  layer_type: other
  args: ...
forward: my_ml_model.jl
```

Then I think all architectures can be supported.

Some benefits:
- only invalid PEtab IDs need to be specified in the mapping table (e.g. Greek symbols in Julia). All other IDs can be taken from the YAML format, just like SBML. This is consistent with the original intention of the PEtab v2 mapping table. Because the model already necessarily has IDs for inputs, we should try to use them, and the net YAML enables this.
- until now, users should anyway use the YAML format when it is sufficient. In the long-term, we might be able to expand the YAML format layer types etc. to support all architectures. Requiring that all ML models are specified in a net YAML file already (with a `forward` script file) enables this long-term vision and prevents divergence or the development of framework-specific formats.
- information like input IDs and input data files can be specified at a single location (the net YAML file `inputs` section) instead of across multiple tables (mapping + hybridization/condition table). e.g.
```yaml
inputs:
- input_id: input1
  value: image.png
- input_id: input2
  value: condition_specific_values.hdf5
```
  - the PEtab problem YAML extension section becomes much simpler. e.g. don't necessarily need to define array files there.

The mapping table is still required, to access individual input or output scalars, unless we also support assignment of IDs to these in the net YAML. But it's fine with me to keep the mapping table for this, since this is what the PEtab v2 mapping table was originally made for.

**Comment by sebapersson** [comment]
In case the neural network is provided in another format (e.g. Lux.jl) I would like to keep it purely in Lux instead of starting to mix (as I think this will be easier for users).

As for inputs, I like the current solution with mapping table, as all neural network entities (inputs, outputs, parameters) are treated in a consistent manner.

Still, would like to hear what @FFroehlich thinks.

**Comment by FFroehlich** [comment]
> > all neural network entities (inputs, outputs, parameters) are treated in a consistent manner
> 
> I disagree, because layers have their IDs defined outside of the mapping table. But overall it's not so bad to keep the non-layer IDs in the mapping table. I just think it's better to only use the mapping table where necessary. Inputs, layers and outputs necessarily have IDs in the model itself, so those should be used if they are valid PEtab IDs IMO.

I agree with this point, but not necessarily with the notion that we want to impose the YAML format on users. I think most users will want to be able to specify layers in pytorch where IDs can be extracted from names of model attributes. I see how Julias focus on functional definition makes this more tricky, but if there is nothing that can easily be serialised with extractable layer identifiers, any model exchange will also be tricky and there is not going to be much use in standardisation.

**Comment by sebapersson** [comment]
So in conclusion, I think not enforce a YAML forward trace. Looking forward, it can be relevant to try to export Lux.jl networks to the YAML format (which I beleive can be doable in certain cases, because can we import, export should be in principle be doable).

Further, if do not impose the YAML format, I think input and outputs should be specified in the mapping table, so that we stay consistent.

**Comment by dilpath** [comment]
> any model exchange will also be tricky and there is not going to be much use in standardisation.

Yes, validation will be impossible. And agreed, if validation is possible, then conversion to YAML must also be possible. If validation is not possible, then there is perhaps no point in the NN model YAML at all.

Alright, then I suggest we aim for enforcement for now.

With the current NN model YAML format, it will be easy for libraries to extract PEtab IDs for inputs and layers, but not outputs, so I suggest an additional outputs section. Here's a suggestion for the enforced NN model YAML format, covering:
- model, input, layer, and output PEtab IDs
- integration of HDF5 files for input data or frozen parameter values
- forward trace or script

```yaml
nn_model_id: nn1

inputs:
- input_id: inputId1
- input_id: inputId2
- ...  # additional inputs

layers:
- layer_id: conv1
  layer_type: Conv2d
  ...  # additional Conv2d properties
- layer_id: lin1
  layer_type: Linear
  ...  # additional Linear properties
- ...  # additional layers

outputs:
- output_id: outputId1
- output_id: outputId2
- ...  # additional outputs

# exactly one of `forward_trace` or `forward_script` are required
forward_trace:
  ...  # the forward trace
forward_script: my_nn_model.jl

# If we enforce the NN model YAML, then these array files can be specified here too,
# similar to how people might directly integrate best parameter estimates into an SBML file.
# But, fine for me if we only allow them in the problem YAML extension section as currently documented,
# since they contain condition-specific information that isn't relevant to the NN model architecture
input_data_files:
- my_inputs_1.hdf5
- my_other_inputs_2.hdf5
# parameter values can be overridden in the PEtab parameters table, to be estimated (?)
parameter_value_files:
- my_parameters_1.hdf5
- my_other_parameters_2.hdf5
```

So it ends up with similar information to an SBML file.

**Comment by sebapersson** [comment]
> model, input, layer, and output PEtab IDs

So this suggests the input argument names should be given in the YAML, not in the mapping table as currently in the spec?

**Comment by dilpath** [comment]
> > model, input, layer, and output PEtab IDs
> 
> So this suggests the input argument names should be given in the YAML, not in the mapping table as currently in the spec?

Yes, because (from above, my opinion):

> Inputs, layers and outputs necessarily have IDs in the model itself, so those should be used if they are valid PEtab IDs IMO.

This would mean we make the same design choice that PEtab v2 does for mechanistic models: it uses IDs directly from SBML or PySB if they are already valid PEtab IDs.

**Comment by sebapersson** [comment]
Ok, but if the model is in for example Lux.jl it is ok to use the mapping table?

**Comment by dilpath** [comment]
Since all IDs will be in the NN model YAML fields (`inputs`, `outputs`, and `layers`), the mapping table would only make sense for:
- invalid PEtab IDs
- individual input/output/parameter scalars (if we allow that)

So these use cases are what we would write in the spec.

**Comment by sebapersson** [comment]
Will paste a comment I made on a PR:

I do not think we should enforce the network YAML as of #17. I believe what has been bothering me is that we are deviating from one of the initial design goals in the docs:  

> The PEtab SciML specification is designed to keep the mechanistic model, NN model, and PEtab problem as independent as possible while linking them through the hybridizatin...

This means we should stop giving special treatment to neural networks in the YAML format. Adding NN-model‐format-specific cases only makes the specification more confusing and leads to having multiple ways to accomplish the same thing, which effort has been put into the spec to avoid up to this point.

**Comment by dilpath** [comment]
> I do not think we should enforce the network YAML as of [#17](https://github.com/sebapersson/petab_sciml/issues/17). I believe what has been bothering me is that we are deviating from one of the initial design goals in the docs:
> 
> > The PEtab SciML specification is designed to keep the mechanistic model, NN model, and PEtab problem as independent as possible while linking them through the hybridizatin...

Enforcing the NN YAML does not reduce the independence of the mechanistic model, NN model, and PEtab problem. Enforcing it will improve the independence, because now NN model properties (like input IDs) are in the NN model YAML, not the PEtab problem mapping table.

> This means we should stop giving special treatment to neural networks in the YAML format. Adding NN-model‐format-specific cases only makes the specification more confusing and leads to having multiple ways to accomplish the same thing, which effort has been put into the spec to avoid up to this point.

How will enforcing the NN model YAML lead to multiple ways of accomplishing the same thing?
- If we don't enforce the NN model YAML, there is one way to specify NN model IDs: the mapping table.
- If we enforce the NN model YAML, there is one way to specify NN model IDs: the NN model YAML. Above I wrote that the mapping table could still be used for mapping scalars, but even this can be handled within the NN model YAML.

> Still, having named input arguments for the forward function can be nice (when applicable), and checked this can also be done in Lux). But if we allow this, for consistency the assignment of the named input to a file should occur in the hybridization or condition table, not inside a YAML input file.

Actually I think it's more consistent if we only use the hybridization table to link the NN with the mechanistic/observation model. Specifying input data tensors for the neural network is not a hybridization. We do not require users to specify the initial condition of their ODE in the hybridization table, so why does this make sense for the input data of the NN?

Enforcing the NN model YAML enables the following design:
- purely mechanistic information in the SBML file
- purely NN information in the NN model YAML
- hybridization information with either the mechanistic or observable models
  - condition-independent: put it in the hybridization table
  - condition-dependent: put it in the condition table

**Comment by sebapersson** [comment]
> Enforcing the NN YAML does not reduce the independence of the mechanistic model, NN model, and PEtab problem. Enforcing it will improve the independence, because now NN model properties (like input IDs) are in the NN model YAML, not the PEtab problem mapping table.

This is a fair point, but I think maybe the best way to address this would be to rather treat the input argument in the YAML (or the Lux.jl and equinox) as valid PEtab Ids if they are on the correct form (otherwise they must be mapped via the mapping table. Because if we enforce the YAML,  I see that we decrease the independence as we kind of force the neural network (regardless of format) into the PEtab problem as a YAML file.

> How will enforcing the NN model YAML lead to multiple ways of accomplishing the same thing?

Sorry, this is probably more related to #20 where with `targetId` in the HDF5 file I felt we were getting multiple ways to map to a file.

> Actually I think it's more consistent if we only use the hybridization table to link the NN with the mechanistic/observation model. Specifying input data tensors for the neural network is not a hybridization. We do not require users to specify the initial condition of their ODE in the hybridization table, so why does this make sense for the input data of the NN?

This is a fair point. But, the hybridization, parameter, or mapping table is how we currently assign inputs for cases when do not have a tensor input. I believe it would be nice to keep this consistent. I feel we are really getting into semantics here :)

**Comment by FFroehlich** [comment]
I think the consensus during the COMBINE session was, pretty consistently, to avoid inventing a new markup format for NNs and the YAML format ends up doing that anyways. I am strongly opposed to making the YAML format mandatory and this should really only be there to facilitate model exchange. Users will want to formulate their models in pytorch etc and we should enable them to do so.

**Comment by sebapersson** [comment]
Given the discussion above, I think we should then do the following to move forward:

1. We do not enforce the YAML network format.
2. Input argument names to the neural network are, if formatted correctly, considered valid PEtab ids. Generally input names can be extracted from a neural network library. If this is not the case, the, the mapping table can always be used.

If everyone agree, we can close this issue, and update #20 accordingly.

**Comment by dilpath** [comment]
> I think the consensus during the COMBINE session was, pretty consistently, to avoid inventing a new markup format for NNs and the YAML format ends up doing that anyways.

The meeting notes only say that the consensus was to use a standard like ONNX. I think the only reason for not inventing a NN format during the COMBINE session was because it would be too much work. But, we did the work, and now the YAML format with external `forward` support is sufficient for PEtab SciML. What were the other reasons for not inventing a new format? We anyway need a format for exchange -- why not expect that all tools (not users!) support reading/writing it?

> I am strongly opposed to making the YAML format mandatory and this should really only be there to facilitate model exchange. Users will want to formulate their models in pytorch etc and we should enable them to do so.

Just to clarify: enforcing the YAML does not mean the user needs to specify their model in YAML. Formulating models in pytorch/Lux.jl is supported regardless of whether we enforce the YAML format. The tool needs to be able to support non-YAML import regardless, because users are never expected to directly write the NN in YAML (like ONNX). Enforcing the YAML format means tools need to be able to read non-YAML models and export to YAML, so that if users e.g. publish their PEtab SciML problem, they need to include the generated YAML file too. The YAML format is an export format for exchange and also means we only need to implement validation once, like petablint. Like with ONNX or SBML, the user never needs to touch the YAML file.

Off-topic: looking back at the HARMONY meeting notes, we also decided to support black-box models, not just neural networks. Fine for me to only support NNs now though. But the YAML format trivially supports black-box models through its external script support.

> If everyone agree, we can close this issue, and update #20 accordingly.

Fine for me! I can update it to stress that the YAML format is for exchange and that it's never required.

**Comment by sebapersson** [comment]
> Off-topic: looking back at the HARMONY meeting notes, we also decided to support black-box models, not just neural networks. Fine for me to only support NNs now though. But the YAML format trivially supports black-box models through its external script support.

I mean we do this implicitly, as in most machine learning frameworks you can code a large family of models (e.g. regression models).

**Comment by dilpath** [comment]
> > Off-topic: looking back at the HARMONY meeting notes, we also decided to support black-box models, not just neural networks. Fine for me to only support NNs now though. But the YAML format trivially supports black-box models through its external script support.
> 
> I mean we do this implicitly, as in most machine learning frameworks you can code a large family of models (e.g. regression models). None differentiable models, e.g. random forest, we probably never want to support though, as we cannot get gradients.

Ah, I only meant to note that the scope of PEtab SciML has changed and is currently only NNs now, after https://github.com/sebapersson/petab_sciml/pull/19#discussion_r2052061959
I didn't mean to say that the YAML has an advantage here over the flexible formats.

**Comment by FFroehlich** [comment]
I am fully on board with enforcing YAML acts as exchange format and allowing specifications in other formats. But that implies that we need lossless conversion between formats, meaning that corresponding IDs for layers/inputs/outputs need to be either inferable from, e.g., a pytorch model or specified in the mapping table.

**Comment by dilpath** [comment]
Yes, the workflow could get messy...

I think inferable IDs are not a problem, since we have the `$nnId.inputs{[$inputArgumentIndex]{[$inputIndex]}}` syntax that we can annotate all ambiguous IDs with.

One implementation option is that the user writes their model in pytorch (or Lux.jl) and uses `petab_sciml` (or PEtab.jl) to generate the YAML from it, which includes some automatically generated mapping with our `$nnId.inputs{[$inputArgumentIndex]{[$inputIndex]}}` syntax, using some default pattern if input IDs cannot be determined.

```yaml
mlmodel_id: model1
inputs:
- input_id: model1_input0
  reference: model1.inputs[0]
- input_id: model1_input1
  reference: model1.inputs[1]
```

The user then writes their PEtab tables with these IDs. But then the YAML becomes required for specification.

The alternative of using the mapping table is fine too but then we are using it to specify IDs, rather than mapping IDs.

Overall I don't see a nice solution to this problem yet.

**Comment by sebapersson** [comment]
> Enforcing the YAML format means tools need to be able to read non-YAML models and export to YAML, so that if users e.g. publish their PEtab SciML problem, they need to include the generated YAML file too. The YAML format is an export format for exchange and also means we only need to implement validation once, like petablint. Like with ONNX or SBML, the user never needs to touch the YAML file.

This I think is to hard to achieve, as I believe some model structures will be non-trivial to translate to the YAML format. But I think tools should be encouraged to be able to export to the YAML format when possible. 

> The mapping table becomes useless if I add inputs to the spec, so I would additionally opt to drop the mapping table.

I do not follow this line of thinking.

**Comment by dilpath** [comment]
> > Enforcing the YAML format means tools need to be able to read non-YAML models and export to YAML, so that if users e.g. publish their PEtab SciML problem, they need to include the generated YAML file too. The YAML format is an export format for exchange and also means we only need to implement validation once, like petablint. Like with ONNX or SBML, the user never needs to touch the YAML file.
> 
> This I think is to hard to achieve, as I believe some model structures will be non-trivial to translate to the YAML format. But I think tools should be encouraged to be able to export to the YAML format when possible.
> 
> > The mapping table becomes useless if I add inputs to the spec, so I would additionally opt to drop the mapping table.
> 
> I do not follow this line of thinking.

Hm, I don't think this is a problem, but maybe some details were lost in the discussion, so I will write up the suggestion comprehensively next week.

**Comment by FFroehlich** [comment]
What we could do is allow the ID nomenclature we talked about above and allow its use directly in the hybridisation table. Is that what you are suggesting? I think this would be fine as we are no longer doing any mapping in the condition tables.

**Comment by sebapersson** [comment]
Following the discussion yesterday I will close this.
